_posts/2019-07-10-feature-columns/feature_columns.Rmd

```diff
@@ -0,0 +1,566 @@
+$ ps_calc_01     <dbl> 0.6, 0.3, 0.5, 0.6, 0.4, 0.7, 0.2, 0.1, 0.9, 0.…
+$ ps_calc_02     <dbl> 0.5, 0.1, 0.7, 0.9, 0.6, 0.8, 0.6, 0.5, 0.8, 0.…
+$ ps_calc_03     <dbl> 0.2, 0.3, 0.1, 0.1, 0.0, 0.4, 0.5, 0.1, 0.6, 0.…
+$ ps_calc_04     <dbl> 3, 2, 2, 2, 2, 3, 2, 1, 3, 2, 2, 2, 4, 2, 3, 2,…
+$ ps_calc_05     <dbl> 1, 1, 2, 4, 2, 1, 2, 2, 1, 2, 3, 2, 1, 1, 1, 1,…
+$ ps_calc_06     <dbl> 10, 9, 9, 7, 6, 8, 8, 7, 7, 8, 8, 8, 8, 10, 8, …
+$ ps_calc_07     <dbl> 1, 5, 1, 1, 3, 2, 1, 1, 3, 2, 2, 2, 4, 1, 2, 5,…
+$ ps_calc_08     <dbl> 10, 8, 8, 8, 10, 11, 8, 6, 9, 9, 9, 10, 11, 8, …
+$ ps_calc_09     <dbl> 1, 1, 2, 4, 2, 3, 3, 1, 4, 1, 4, 1, 1, 3, 3, 2,…
+$ ps_calc_10     <dbl> 5, 7, 7, 2, 12, 8, 10, 13, 11, 11, 7, 8, 9, 8, …
+$ ps_calc_11     <dbl> 9, 3, 4, 2, 3, 4, 3, 7, 4, 3, 6, 9, 6, 2, 4, 5,…
+$ ps_calc_12     <dbl> 1, 1, 2, 2, 1, 2, 0, 1, 2, 5, 3, 2, 3, 0, 1, 2,…
+$ ps_calc_13     <dbl> 5, 1, 7, 4, 1, 0, 0, 3, 1, 0, 3, 1, 3, 4, 3, 6,…
+$ ps_calc_14     <dbl> 8, 9, 7, 9, 3, 9, 10, 6, 5, 6, 6, 10, 8, 3, 9, …
+$ ps_calc_15_bin <int> 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0,…
+$ ps_calc_16_bin <int> 1, 1, 1, 0, 0, 1, 1, 0, 1, 1, 1, 1, 0, 0, 0, 1,…
+$ ps_calc_17_bin <int> 1, 1, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 1, 0, 1,…
+$ ps_calc_18_bin <int> 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0,…
+$ ps_calc_19_bin <int> 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1,…
+$ ps_calc_20_bin <int> 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0,…
+```
+
+We split off 25% for validation.
+
+```{r}
+# train-test split
+id_training <- sample.int(nrow(porto), size = 0.75*nrow(porto))
+
+x_train <- porto[id_training,] %>% select(-target)
+x_test <- porto[-id_training,] %>% select(-target)
+y_train <- porto[id_training, "target"]
+y_test <- porto[-id_training, "target"] 
+```
+
+The only thing we want to do to the _data_ before defining the network is scaling the numeric features. Binary and categorical features can stay as is, with the minor correction that for the categorical ones, we'll actually pass the network the numeric representation of the factor data.
+
+Here is the scaling.
+
+```{r}
+train_means <- colMeans(x_train[sapply(x_train, is.double)]) %>% unname()
+train_sds <- apply(x_train[sapply(x_train, is.double)], 2, sd)  %>% unname()
+train_sds[train_sds == 0] <- 0.000001
+
+x_train[sapply(x_train, is.double)] <- sweep(
+  x_train[sapply(x_train, is.double)],
+  2,
+  train_means
+  ) %>%
+  sweep(2, train_sds, "/")
+x_test[sapply(x_test, is.double)] <- sweep(
+  x_test[sapply(x_test, is.double)],
+  2,
+  train_means
+  ) %>%
+  sweep(2, train_sds, "/")
+```
+
+When building the network, we need to specify the input and output dimensionalities for the embedding layers. Input dimensionality refers to the number of different symbols that "come in"; in NLP tasks this would be the vocabulary size while here, it's simply the number of values a variable can take. ^[In this dataset, every categorical variable is a column of singleton integers.] 
```

Comment:
```suggestion
When building the network, we need to specify the input and output dimensionalities for the embedding layers. _Input dimensionality_ refers to the number of different symbols that "come in." In NLP tasks, this would be the vocabulary size while here, it's simply the number of values a variable can take. ^[In this dataset, every categorical variable is a column of singleton integers.] 
```

Reply:
Or some form of emphasis when you give a nice, pithy definition!